README.md
# Fauxhai-ng

![CI](https://github.com/chef/fauxhai/workflows/CI/badge.svg) [![Gem Version](https://badge.fury.io/rb/fauxhai-ng.svg)](https://badge.fury.io/rb/fauxhai-ng)

Note: fauxhai-ng is an updated version of the original fauxhai gem. The CLI and library namespaces have not changed, but you will want to update to use the new gem.

Fauxhai is a gem for mocking out [ohai](https://github.com/chef/ohai) data in your chef testing. Fauxhai is community supported, so we need **your help** to populate our dataset. Here's an example for testing my "awesome_cookbook" on Ubuntu 20.04:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.mock(platform: 'ubuntu', version: '20.04')
  end

  it 'should install awesome' do
    @runner = ChefSpec::ChefRunner.new.converge('tmpreaper::default')
    expect(@runner).to install_package 'awesome'
  end
end
```

Alternatively, you can pull "real" Ohai data from an existing server:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.fetch(host: 'server01.example.com')
  end

  it 'should install awesome' do
    @runner = ChefSpec::ChefRunner.new.converge('tmpreaper::default')
    expect(@runner).to install_package 'awesome'
  end
end
```

## Important Note

Fauxhai ships with a command line tool - `fauxhai`. This is **not** the same as Fauxhai.mock. Running `fauxhai` on a machine effectively runs `ohai`, but then sanitizes the data, removing/replacing things like:

- users
- ssh keys
- usernames in paths
- sensitive system information

`fauxhai` should only be used by developers wishing to submit a new json file.

## Platform and Versions

For a complete list of platforms and versions available for mocking via Fauxhai see [PLATFORMS.MD](https://github.com/chef/fauxhai/blob/master/PLATFORMS.md) in this repository.

## Usage

Fauxhai provides a bunch of default attributes so that you don't need to mock out your entire infrastructure to write a simple test. That being said, not all configurations will suit your needs. Because of that, Fauxhai provides two ways to configure your mocks:

### Overriding

`Fauxhai.mock` will also accept a block with override attributes that are merged with all the default attributes. For example, the default Ubuntu 12.04 mock uses `Ruby 1.9.3`. Maybe your system is using `ree`, and you want to verify that the cookbooks work with that data as well:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.mock(platform: 'ubuntu', version: '12.04') do |node|
      node['languages']['ruby']['version'] = 'ree'
    end
  end

  it 'should install awesome' do
    @runner = ChefSpec::ChefRunner.new.converge('tmpreaper::default')
    expect(@runner).to install_package 'awesome'
  end
end
```

The `node` block variable allows you to set any Ohai attribute on the mock that you want. This provides an easy way to manage your environments. If you find that you are overriding attributes like OS or platform, you should see the section on Contributing.

### Fetching

Alternatively, if you do not want to mock the data, Fauxhai provides a `fetch` mechanism for collecting "real" Ohai data from a remote server or local file. Maybe you want to test against the fully-replicated environment for a front-facing server in your pool. Just pass in the `url` option instead of a `platform`:

The `fetch` method supports all the same options as the Net-SSH command, such as `:user`, `:password`, `:key_file`, etc.

The `fetch` method will cache the JSON file in a temporary path on your local machine. Similar to gems like VCR, this allows Fauxhai to use the cached copy, making your test suite run faster. You can optionally force a cache miss by passing the `:force_cache_miss => true` option to the `fetch` initializer. **Because this is real data, there may be a security concern. Secure your laptop accordingly.**

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.fetch(host: 'server01.example.com')
  end

  it 'should install awesome' do
    @runner = ChefSpec::ChefRunner.new.converge('tmpreaper::default')
    expect(@runner).to install_package 'awesome'
  end
end
```

This will ssh into the machine (you must have authorization to run `sudo ohai` on that machine), download a copy of the Ohai output, and optionally cache that data inside the test directory (speeding up future tests).

### Overriding + Fetching

As you might expect, you can combine overriding and fetching like so:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.fetch(host: 'server01.example.com') do |node|
      node['languages']['ruby']['version'] = 'ree'
    end
  end

  it 'should install awesome' do
    @runner = ChefSpec::ChefRunner.new.converge('tmpreaper::default')
    expect(@runner).to install_package 'awesome'
  end
end
```

### Fixturing

If you want to use Fauxhai as "fixture" data, you can store real JSON in your project and use the `:path` option:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.mock(path: 'fixtures/my_node.json')
  end
end
```

### Overriding + Fixturing

You can also change specific attributes in your fixture:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.mock(path: 'fixtures/my_node.json') do |node|
      node['languages']['ruby']['version'] = 'ree'
    end
  end
end
```

### Disabling Fetching from Github

On environments that does not have access to the internet, you can disable fetching Fauxhai data from GitHub as follow:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  before do
    Fauxhai.mock(platform: 'ubuntu', version: '12.04', github_fetching: false)
  end
end
```

## Testing Multiple Versions

It's a common use case to test multiple version of the same operating system. Here's a simple example to get your started. This is more rspec-related that Fauxhai related, but here ya go:

```ruby
require 'chefspec'

describe 'awesome_cookbook::default' do
  ['18.04', '20.04'].each do |version|
    context "on Ubuntu #{version}" do
      before do
        Fauxhai.mock(platform: 'ubuntu', version: version)
      end

      it 'should install awesome' do
        @runner = ChefSpec::ChefRunner.new.converge('tmpreaper::default')
        expect(@runner).to install_package 'awesome'
      end
    end
  end
end
```

## Contributing

See [CONTRIBUTING.md](https://github.com/chef/fauxhai/blob/master/CONTRIBUTING.md).


CHANGELOG.md
<!-- latest_release 9.3.13 -->
## [v9.3.13](https://github.com/chef/fauxhai/tree/v9.3.13) (2023-12-05)

#### Merged Pull Requests
- Added WIndows 2022 and Updated CONTRIBUTING.md [#24](https://github.com/chef/fauxhai/pull/24) ([jjperry](https://github.com/jjperry))
<!-- latest_release -->

<!-- release_rollup since=9.3.8 -->
### Changes not yet released to rubygems.org

#### Merged Pull Requests
- Added WIndows 2022 and Updated CONTRIBUTING.md [#24](https://github.com/chef/fauxhai/pull/24) ([jjperry](https://github.com/jjperry)) <!-- 9.3.13 -->
- Added RedHat 9 [#21](https://github.com/chef/fauxhai/pull/21) ([jjperry](https://github.com/jjperry)) <!-- 9.3.12 -->
- Add Ubuntu 22.04 (jammy) support [#16](https://github.com/chef/fauxhai/pull/16) ([dtseiler](https://github.com/dtseiler)) <!-- 9.3.11 -->
- Add .buildkite folder to make pipeline_protection happy [#18](https://github.com/chef/fauxhai/pull/18) ([tpowell-progress](https://github.com/tpowell-progress)) <!-- 9.3.10 -->
- Add Debian 12 support [#17](https://github.com/chef/fauxhai/pull/17) ([Justin-Fernbaugh](https://github.com/Justin-Fernbaugh)) <!-- 9.3.9 -->
<!-- release_rollup -->

<!-- latest_stable_release -->
## [v9.3.8](https://github.com/chef/fauxhai/tree/v9.3.8) (2023-05-05)

#### Merged Pull Requests
- Add Rocky Linux 9 data [#9](https://github.com/chef/fauxhai/pull/9) ([vkarve-chef](https://github.com/vkarve-chef))
- Added the SonarQube configuration [#13](https://github.com/chef/fauxhai/pull/13) ([nikhil2611](https://github.com/nikhil2611))
- Update fetcher to pull from correct repository [#14](https://github.com/chef/fauxhai/pull/14) ([ramereth](https://github.com/ramereth))
<!-- latest_stable_release -->

## [v9.3.5](https://github.com/chef/fauxhai/tree/v9.3.5) (2023-02-03)

#### Merged Pull Requests
- Add AlmaLinux 9 data [#8](https://github.com/chef/fauxhai/pull/8) ([vkarve-chef](https://github.com/vkarve-chef))
- CentOS Stream 8 &amp; 9 data [#10](https://github.com/chef/fauxhai/pull/10) ([ramereth](https://github.com/ramereth))
- Add Oracle 7 / 8 / 9 [#12](https://github.com/chef/fauxhai/pull/12) ([davidjkling](https://github.com/davidjkling))

## [v9.3.2](https://github.com/chef/fauxhai/tree/v9.3.2) (2022-06-13)

#### Merged Pull Requests
- Mimaslanka/dummy [#6](https://github.com/chef/fauxhai/pull/6) ([mimaslanka](https://github.com/mimaslanka))
- Force action [#7](https://github.com/chef/fauxhai/pull/7) ([mimaslanka](https://github.com/mimaslanka))

## 9.3.0 (2022-01-26)

- Add macOS 12
- Deprecate macOS 11.0 with a new macOS 11 taking it's place
- Deprecate FreeBSD 11.2 since this is EOL now

## 9.2.0 (2022-01-26)

- Add Alma Linux 8
- Add Rocky Linux 8

## 9.1.0 (2021-11-01)

### Updated Platforms

- Debian 10 updated to 10.11
- Debian 11 regenerated on AWS using Debian 11.1
- Suse 15 updated from 15.0 to 15.1
- Amazon 2 updated to tha latest release
- Ubuntu 18.04 / 20.04 updated to the latest releases
- Redhat 8 updated from 8.2 to 8.4

### New Platforms

- Debian 9.13
- Redhat 7.9

### Newly Deprecated Platforms (removal in April 2022)

- Redhat 7.7
- CentOS 7.7
- Debian 9.11
- Fedora 31
- Oracle 7.5
- Amazon 2018.03

## v9.0.1 (2021-11-01)

- Resolved a Ruby error when attempting to generate new Fauxhai data on a host
- Deprecated the SmartOS platform data dump
- Add Debian 11

## v9.0.0 (2021-04-07)

This release removes all previously deprecated Fauxhai definitions which have alerted during their use over the last year.  These platforms have been removed either due to the platform reaching end-of-life status with the vendor or newer minor releases of the platform being available in Fauxhai.

In general make sure to write ChefSpec tests that reference the major version (or no version at all) of the platform so that Fauxhai can automatically pick the latest version. Allowing Fauxhai to best chose the version of the minor or major platform release ensures your specs continue to run as new platforms are released and older Fauxhai data is removed.

To automatically migrate ChefSpec tests to use newer platforms you use the [Chef/Deprecations/DeprecatedChefSpecPlatform](https://docs.chef.io/workstation/cookstyle/chef_deprecations_deprecatedchefspecplatform/) Cookstyle cop, which will autocorrect many specs:

`cookstyle -a --only Chef/Deprecations/DeprecatedChefSpecPlatform`

### Platforms Removed

- AIX 6.1
- Amazon Linux 2015.03
- Amazon Linux 2015.09
- Amazon Linux 2016.03
- Amazon Linux 2016.09
- Amazon Linux 2017.03
- Amazon Linux 2017.09
- CentOS 5.11
- CentOS 6.9
- CentOS 7.5.1804
- CentOS 7.6.1810
- Debian 8.11
- Debian 9.8
- Debian 9.9
- Fedora 29
- Fedora 30
- FreeBSD  11.1
- Linux Mint 18.3
- macOS 10.13
- OpenSUSE 15.0
- OpenSUSE 15.1
- Oracle Linux 5.11
- Oracle Linux 6.9
- Red Hat Linux 5.11
- Red Hat Linux 6.9
- Red Hat Linux 7.4
- Red Hat Linux 7.5
- Red Hat Linux 7.6
- Suse Enterprise Linux 12.3
- Ubuntu 14.04
- Windows 20008 R2

## v8.7.1 (2021-04-03)

- Add files to the gem package to help with Chef Workstation bundling

## v8.7.0 (2020-12-31)

### Newly Deprecated Platforms (removal in April 2021)

- openSUSE 15.1
- Redhat 7.6
- CentOS 7.6
- Linux Mint 18.3
- All Amazon Linux 201X releases except 2018.03

## v8.6.0 (2020-12-01)

### New Platforms

- macOS 11

### Newly Deprecated Platforms (removal in April 2021)

- macOS 10.13

## v8.5.0 (2020-11-17)

### New Platforms

- AIX 7.2

### Updated Platforms

- AIX 7.1: Updated the Ohai dump from Chef Infra Client 14 to 16.7, which updates Ruby plugin data, filesystem plugin data, and shard plugin data.

## v8.4.0 (2020-11-09)

### New Platforms

- CentOS 7.8
- Red Hat 7.8
- Add openSUSE 15.2

### Updated Platforms

- Ubuntu 16.04 - 20.04
- Debian 9 & 10 releases
- Red Hat 7 and 8 releases
- openSUSE 15.1
- Fedora 31/32
- Amazon releases

## v8.3.1 (2020-08-17)

- Only require net/ssh if we're using the ssh functionality
- Add a rake task to generate a platforms.json file that can be parses out to determine all platforms

## v8.3.0 (2020-08-13)


lib/fauxhai/runner/default.rb
module Fauxhai
  class Runner
    module Default
      def bin_dir
        "/usr/local/bin"
      end

      def counters
        {
          "network" => {
            "interfaces" => {
              "lo" => {
                "tx" => {
                  "queuelen" => "1",
                  "bytes" =>  0,
                  "packets" =>  0,
                  "errors" =>  0,
                  "drop" =>  0,
                  "carrier" => 0,
                  "collisions" => 0,
                },
                "rx" =>  {
                  "bytes" => 0,
                  "packets" => 0,
                  "errors" => 0,
                  "drop" => 0,
                  "overrun" => 0,
                },
              },
              default_interface.to_s => {
                "rx" => {
                  "bytes" => 0,
                  "packets" => 0,
                  "errors" => 0,
                  "drop" => 0,
                  "overrun" => 0,
                  "frame" => 0,
                  "compressed" => 0,
                  "multicast" => 0,
                },
                "tx" => {
                  "bytes" => 0,
                  "packets" => 0,
                  "errors" => 0,
                  "drop" => 0,
                  "overrun" => 0,
                  "collisions" => 0,
                  "carrier" => 0,
                  "compressed" => 0,
                },
              },
            },
          },
        }
      end

      def current_user
        "fauxhai"
      end

      def default_gateway
        "10.0.0.1"
      end

      def default_interface
        case @system.data["platform_family"]
        when "mac_os_x"
          "en0"
        when /bsd/
          "em0"
        when "arch", "fedora"
          "enp0s3"
        else
          "eth0"
        end
      end

      def domain
        "local"
      end

      def fqdn
        "fauxhai.local"
      end

      def gem_bin
        "/usr/local/bin/gem"
      end

      def gems_dir
        "/usr/local/gems"
      end

      def hostname
        "Fauxhai"
      end

      def ipaddress
        "10.0.0.2"
      end

      def ip6address
        "fe80:0:0:0:0:0:a00:2"
      end

      def keys
        {
          "ssh" => ssh,
        }
      end

      def languages
        {
          "ruby" => @system.data["languages"]["ruby"].merge("bin_dir" => bin_dir,
                                                            "gem_bin" => gem_bin,
                                                            "gems_dir" => gems_dir,
                                                            "ruby_bin" => ruby_bin),
          "powershell" => @system.data["languages"]["powershell"],
        }
      end

      def macaddress
        "11:11:11:11:11:11"
      end

      def network
        {
          "interfaces" => {
            "lo" => {
              "mtu" => "65536",
              "flags" => %w{LOOPBACK UP LOWER_UP},
              "encapsulation" => "Loopback",
              "addresses" => {
                "127.0.0.1" => {
                  "family" => "inet",
                  "prefixlen" => "8",
                  "netmask" => "255.0.0.0",
                  "scope" => "Node",
                  "ip_scope" => "LOOPBACK",
                },
                "::1" => {
                  "family" => "inet6",
                  "prefixlen" => "128",
                  "scope" => "Node",
                  "tags" => [],
                  "ip_scope" => "LINK LOCAL LOOPBACK",
                },
              },
              "state" => "unknown",
            },
            default_interface.to_s => {
              "type" => default_interface.chop,
              "number" => "0",
              "mtu" => "1500",
              "flags" => %w{BROADCAST MULTICAST UP LOWER_UP},
              "encapsulation" => "Ethernet",
              "addresses" => {
                macaddress.to_s => {
                  "family" => "lladdr",
                },
                ipaddress.to_s => {
                  "family" => "inet",
                  "prefixlen" => "24",
                  "netmask" => "255.255.255.0",
                  "broadcast" => "10.0.0.255",
                  "scope" => "Global",
                  "ip_scope" => "RFC1918 PRIVATE",
                },
                "fe80::11:1111:1111:1111" => {
                  "family" => "inet6",
                  "prefixlen" => "64",
                  "scope" => "Link",
                  "tags" => [],
                  "ip_scope" => "LINK LOCAL UNICAST",
                },
              },
              "state" => "up",
              "arp" => {
                "10.0.0.1" => "fe:ff:ff:ff:ff:ff",
              },
              "routes" => [
                {
                  "destination" => "default",
                  "family" => "inet",
                  "via" => default_gateway,
                },
                {
                  "destination" => "10.0.0.0/24",
                  "family" => "inet",
                  "scope" => "link",
                  "proto" => "kernel",
                  "src" => ipaddress,
                },
                {
                  "destination" => "fe80::/64",
                  "family" => "inet6",
                  "metric" => "256",
                  "proto" => "kernel",
                },
              ],
              "ring_params" => {},
            },
          },
          "default_interface" => default_interface,
          "default_gateway" => default_gateway,
        }
      end

      def ruby_bin
        "/usr/local/bin/ruby"
      end

      def ssh
        {
          "host_dsa_public" => File.read(File.join(Fauxhai.root, "lib", "fauxhai", "keys", "id_dsa.pub")).strip,
          "host_rsa_public" => File.read(File.join(Fauxhai.root, "lib", "fauxhai", "keys", "id_rsa.pub")).strip,
        }
      end

      def uptime
        "30 days 15 hours 07 minutes 30 seconds"
      end

      def uptime_seconds
        2646450
      end

      def cpu
        {
          "real" => 1,
          "total" => 1,
          "cores" => 1,
        }
      end

      def memory
        {
          "total" => "1048576kB",
        }
      end

      def virtualization
        {
          "systems" => {},
        }
      end

      def time
        {
          "timezone" => "GMT",
        }
      end

      # Whitelist attributes are attributes that we *actually* want from the node. Other attributes are
      # either ignored or overridden, but we ensure these are returned with the command.
      #
      # @return [Array] - the key of whitelisted attributes
      def whitelist_attributes
        %w{
          block_device
          chef_packages
          command
          dmi
          filesystem
          fips
          init_package
          kernel
          lsb
          ohai_time
          os
          os_release
          os_version
          packages
          platform
          platform_version
          platform_build
          platform_family
          root_group
          shard_seed
          shells
        }
      end
    end
  end
end
